Implements a way to adapt the number of bins used for forgetting curve graphs

JSchoreels opened the pull request:
1. Implemented the bucket method for long term forgetting curve
2. Added a way to select a custom amount of bins directly in the graph

https://github.com/user-attachments/assets/d43ec96f-9607-4092-a6ce-6429575c6750

cc @L-M-Sherlock

JSchoreels commented:
Also added a Toggle to filter or not outliers.

For short term reviews I didn't add it because otherwise it would lead to no data

https://github.com/user-attachments/assets/2dcba5b0-5d6a-4005-8651-66ccb73b7ffe

L-M-Sherlock commented:
I'm OK with the new name. I named it like that because SuperMemo also names it like that.

L-M-Sherlock reviewed (approved):
LGTM

Luc-Mcgrady commented:
I have made some changes to the appearance of the inputs. Are you happy for me to merge this still? 

As for the title, I like it as it is because it's already been published under that name so people might wonder if the graph content changed with the title.

JSchoreels commented:
I tested the changes, on some resolution (1728x1117) it gets me something like this :
![CleanShot 2025-11-10 at 10 08 17@2x|200px](https://github.com/user-attachments/assets/d83e38ff-1453-4822-8279-97b3a528b1a9)


But for most others resolution it's fine ! So I guess good to me.

For the title, I kinda still believe that "Forgetting Curve" is wrong since it's absolutely not a forgetting curve; it's the distribution of Ratings for a specific Interval, but some of those dots could be reviews made after 3-4 reviews. Also, it means that if FSRS has a perfect R accuracy, then the theoretical graph should be a horizontal line, not a forgetting curve.

But at the same time, the fact a forgetting curve is drawn on that graph reinforce that confusion.
I think the portion of the graph with low interval reviews looks like the forgetting curve simply because they are mostly 1-review grades; thus, they kind of look like a legit forgetting curve.

Maybe check on your side how your Good grade looks on the long-term one, but I would not be surprised if it looks something like : 
<img width="1051" height="412" alt="image" src="https://github.com/user-attachments/assets/173d9683-39b4-4754-805d-9fbff70101eb" />
(The increasing curve was manually drawn my @Expertium I think)

I'd love to say I'm not against keeping it as it is, since it's already valuable information, but I think it's better to make it clear enough right away instead of accumulating user confusion about their ascending forgetting curve before having to change it 😆 .

JSchoreels commented:
I mean I don't remember having reviews with good first review planned at ~300d interval (cf bottom right graph)

![CleanShot 2025-11-10 at 10 29 51@2x](https://github.com/user-attachments/assets/d1908ab4-3fa3-45df-875a-fd00c194a0eb)

Or maybe they were reviews made with older params that would have been 300d with newer ones ?

Then I'd be wrong, but there are also some confusing aspects with selecting "Last 12 months" vs "All History".

The params of that specific deck is `0.1189, 0.8616, 2.3040, 23.4339, 6.5891, 0.3549, 2.7763, 0.0053, 1.5089, 0.2041, 0.5716, 1.5512, 0.0187, 0.4488, 1.2714, 0.1580, 3.2025, 0.7075, 0.2425, 0.1270, 0.1494`, so I don't really understand how "Good" first grade could have been >250 interval.

Or maybe it's the "12 last months" that is bugged? And the "All History" being the only one to be looked at ?

I might be wrong on my previous assumptions; maybe it's really the first grade, but another bug makes it look like it's not?

JSchoreels commented:
Looking at 

```
            const first_day = first_rating_day[revlog.cid]
            if (first_day !== undefined && !recorded_cards.has(revlog.cid)) {
                const delta_t = day - first_day
                if (delta_t > 0) {
                    forgetting_samples.push({
                        cid: revlog.cid,
                        firstRating: first_rating[revlog.cid]!,
                        delta: delta_t,
                        recall: revlog.ease > 1 ? 1 : 0,
                    })
                    recorded_cards.add(revlog.cid)
                }
            }

``` 

I have the feeling that the problem is that the first_rating_day, for cards that were reviewed in the past 12 months, are not the very first review but their first review in that 12-month scope. Thus why the graph domain is so big when you select "Last 12 Months)".

I think a easy fix is simply to put the same "Increase Date Rate" for that graph, so users can't see if it they didn't select "All history"

![CleanShot 2025-11-10 at 10 44 02@2x](https://github.com/user-attachments/assets/101b5bd9-49c0-4c5e-a845-b273fa3ed30b)

What do you think @Luc-Mcgrady ?

L-M-Sherlock commented:
> is that so? Then it 100% needs to be changed.

I guess it's just caused by a couple of mixed presets with different FSRS parameters.

JSchoreels commented:
> I recommend only drawing the forgetting curve based on cards introduced in the last 12 months when the 12-month scope is enabled.

That's also a good approach but I'm not sure 100% we have that info ? I checked a bit the code but it seems the `introduced : Set<number>()` might not also be 100% accurate since I see it's filled once we have a review with ivl != 0, cf

```
        } else if (!introduced.has(revlog.cid) && revlog.ivl != 0) {
            introduced_day_count[day] = (introduced_day_count[day] ?? 0) + 1
            [...]
            introduced.add(revlog.cid)
            reintroduced.add(revlog.cid)
            forgotten.delete(revlog.cid)
        }

```  

In any case, I've made 2 changes :
- @Luc-Mcgrady proposal to add `revlog.type == 0` 
- Added the check for memorised data and if not, just show the `MemorisedCalculator`. Might not be ideal because the memorised data are not mandatory for this graph, but at least it helps already (and I don't think doing the memorized calculation is a problem any way if you're already in FSRS graphs ?)

Doing so, now both graphs looks pretty similar so some accuracy might indeed be lost (look at the Again forgetting curve, it's slightly lower in one), but at least it's still makes sense. :) 

<img width="1354" height="632" alt="image" src="https://github.com/user-attachments/assets/c62f71b4-3817-41ac-8541-e98c41a1fe56" />

Luc-Mcgrady commented on src/ts/RevlogGraphs.svelte:
```diff
@@ -552,21 +545,35 @@
+            <p>{i18n("first-short-term-forgetting-curve-help")}</p>
+            {#if truncated}
+                <Warning>{i18n("generic-truncated-warning")}</Warning>
+            {/if}
```
You still used the memorised calculator? Is this intentional?

JSchoreels replied:
Can you check again ? I've simply removed those things since the accuracy isn't that different anyway
